lib/puppetserver/ca/action/list.rb

```diff
@@ -79,16 +89,55 @@ def run(input)
           requested, signed, revoked = separate_certs(all_certs)
           missing = certnames - all_certs.map { |cert| cert['name'] }
 
-          (all || certnames.any?) \
-            ? output_certs_by_state(requested, signed, revoked, missing)
-            : output_certs_by_state(requested)
+          (all || certnames.any?)\
+          ? output_certs_by_state(all, output_format, requested, signed, revoked, missing)
+          : output_certs_by_state(all, output_format, requested)
 
           return missing.any? \
             ? 1
             : 0
         end
+        
+        def output_certs_by_state(all, output_format, requested, signed = [], revoked = [], missing = [])
+          if output_format == 'json'
+            output_certs_json_format(all, requested, signed, revoked, missing)
+          else
+            output_certs_text_format(requested, signed, revoked, missing)
+          end
+        end
 
-        def output_certs_by_state(requested, signed = [], revoked = [], missing = [])
+        def output_certs_json_format(all, requested, signed, revoked, missing)
+          if revoked.empty? && signed.empty? && requested.empty? && missing.empty?
+            @logger.inform([].to_json)
+            return
+          end
+
+          if all
+            grouped_cert = { "requested" => requested,
+                             "signed" => signed,
+                             "revoked" => revoked,
+                             "missing" => missing }.to_json
```

Comment:
Talked with Tu in chat about this already, but for review purposes, "missing" isn't a valid key in `--all` behavior.